File: C++/trayectoria.cpp
#include "trayectoria.h"
#include "tiempo.h"

float x;
float y;
const double PI = 3.141592; 

Trayectoria::Trayectoria(){}
Trayectoria::~Trayectoria(){}

bool Trayectoria::esOrdenValida(string x,Trayectoria * trayec){
//Casos de orden invalida:
//no comienza con E o termina con P
//hay caracteres invalidos
//el siguiente caracter de A,B,C es distinto de +,-
//el siguiente caracte de +,- es distinto de un nro
  for(int i=0;i<x.length();i++){
    if ((x[0]!='E' || x[x.length()-1]!='P')
      || ((i!=0 && i!=x.length()-1)&&(x[i]!='A' && x[i]!='B' && x[i]!='C' 
          && x[i]!='+' && x[i]!='-' && (x[i]-'0'>9 || x[i]-'0'<0)))
      || ((x[i-1]=='A' || x[i-1]=='B' ||x[i-1]=='C')&&(x[i]!='-' && x[i]!='+'))
      || ((x[i-1]=='+' || x[i-1]=='-')&&(x[i]-'0'>9 || x[i]-'0'<0)))
      return false; 
  }
  trayec->guardarAngulo(x);
  //Tiempo una vez ya guardado los ángulos
  Tiempo::medirTime();
  return true;
}

void Trayectoria::guardarAngulo(string x,Trayectoria * trayec){
  int index;
  for(int i=1;i<x.length()-1;i++){
    string aux;
    switch(int(x[i])){
      case(65):
      case(66):
      case(67):
        index=i;
        while(x[i+1]!='A' && x[i+1]!='B'&& x[i+1]!='C'&& x[i+1]!='P'){
          aux += x[i+1];
          i++;
        }
        int angle;
        istringstream(aux)>>angle;

        if(x[index]=='A') anguloGiro[0]=angle;
        if(x[index]=='B') anguloGiro[1]=angle;
        if(x[index]=='C') anguloGiro[2]=angle;

        
        Tiempo::setActivityTime(abs(angle)*0.01);
        Tiempo::medirTime();
        posicion(anguloGiro[0],anguloGiro[1],anguloGiro[2]);
      break;
    }
  }
}

double Trayectoria::setPosicionx(int angulo,Trayectoria * trayec){
    return (cos(angulo*PI/180));
}

double Trayectoria::setPosiciony(int angulo,Trayectoria * trayec){
    return (cos(angulo*PI/180));
}

double Trayectoria::setPosicionz(int angulo,Trayectoria * trayec){
    return (cos(angulo*PI/180));
}

void Trayectoria::posicion(int a, int b, int c,Trayectoria * trayec){
    //Longitudes de cada brazo
    double d1=10,d2=5,d3=4.5;
    //Posición Brazo A en el plano.
    double p1x=0;
    double p1y=0;
    double p1z=0;
    //Posición Brazo B en el plano.
    double p2x=d2*(Trayectoria::setPosicionx(b));
    double p2y=d2*(Trayectoria::setPosiciony(b));
    double p2z=d2*(Trayectoria::setPosicionz(b));
    //Posición Brazo C en el plano.
    double p3x=d3*(Trayectoria::setPosicionx(c)) + p2x;
    double p3y=d3*(Trayectoria::setPosiciony(c)) + p2y;
    double p3z=d3*(Trayectoria::setPosicionz(c)) + p2z;
}

void Trayectoria::setEstado(bool estado,Trayectoria * trayec){
  this->estado = estado; //1-conectado 0-desconectado
}

bool Trayectoria::getEstado(Trayectoria * trayec){
  return this->estado; //1-activo 0-inactivo
}

void Trayectoria::setConexion(bool conexion,Trayectoria * trayec){
  this->conexion = conexion;
}

bool Trayectoria::getConexion(Trayectoria * trayec){
  return conexion;
}

File: POO/poo/trayectoria.py
import sys
sys.path.insert(0, 'home/agustin/Proyecto_poo/Proyecto_poo/POO')
sys.path.insert(0, 'media/datos/UNCUYO/POO/PROYECTO/Proyecto_poo/POO')
import robot
import math
import tiempo as Tiempo
import para_gotod as rgodot

class Trayectoria:
    def __init__(self):
        self.anguloGiro[0]=0
        self.anguloGiro[1]=0
        self.anguloGiro[2]=0  
        self.posicion_artB[3] 
        self.posicion_artC[3] 
        self.orden='EA+50B-130C+3P'
#print("Este es el numero de diferencia entre indices B y A" )
        #print((orden.index('B')-1)-orden.index('A'))    
    
    def guardarAngulo(self, orden):
                #Angulos y signos de cada orden
        self.anguloGiro[0]=int (orden[(orden.index('A')+1):(orden.index('B'))])
        print("angulo 1 ", self.anguloGiro[0])
        self.anguloGiro[1]=int (orden[(orden.index('B')+1):(orden.index('C'))])
        print("angulo 2 ", self.anguloGiro[1])
        self.anguloGiro[2]=int (orden[(orden.index('C')+1):(orden.index('P'))])
        print("angulo 3 ", self.anguloGiro[2])

            #archivo_godot=input("Ingrese un no")
        rgodot.generarOrden(self.anguloGiro[0], self.anguloGiro[1], self.anguloGiro[2])


        Tiempo.setActivityTime(self.anguloGiro[0]*0.01)
        Tiempo.setActivityTime(self.anguloGiro[1]*0.01)
        Tiempo.setActivityTime(self.anguloGiro[2]*0.01)
             
        Tiempo.medirTime()
        posicion(self.anguloGiro[0],self.anguloGiro[1],self.anguloGiro[2])
            
            #*****************************ACA ASIGNAR LOS ANGULOS A LAS CLASES CREADAS**********************************
            #*****************************Mirar de sacar la velocidad y despues ponerla de nuevo************************
            #**********Hacer atributos a robot con todos los angulos y velocidades (EN REALIDAD ASIGNAR UNO A CADA OBJETO QUE SERÍA EL BRAZO)
        velocidad = input("Ingrese la velocidad ")      #En realidad es un argumento pasado por el cliente

    def esOrdenValida(self, x):
        valido= False
        if (x[0]=='E' and x[len(x)-1]=='P'):
            print("es cierto que hay e o p")
            if ('A' in x or 'B' in x or 'C' in x):
                if ((x[x.index('A')+1]=='+' or x[x.index('A')+1]=='-')  and (x[x.index('B')+1]=='+' or x[x.index('B')+1]=='-') and (x[x.index('C')+1]=='+' or x[x.index('C')+1]=='-')):
                    print("Hay un signo de + o de -")
                    if ((1<=abs((x.index('A')+1)-x.index('B'))<=4) and (1<=abs((x.index('B')+1)-x.index('C'))<=4) and (1<=abs((x.index('C')+1)-x.index('P'))<=4)):
                            # or (1<=abs(x.index('A')-x.index('C'))<=4 and (1<=(x.index('B')-x.index('C'))<=4)): # 
                        print("anda")
                            #Llamar a la funcion para variar el ángulo
                            #Las tres variables de abajo probablemente no hagan falta
                        funcion=Trayectoria()
                        funcion.guardarAngulo(x)
                        valido=True
                        return (valido)
                    else:
                        print("Comando inválido") 
                        return(valido)            
                            
                            
                            #Probar de usar Catch errors          
        #diferencias entre A y B y A y C y B y C
        #luego hacer que si las posicions +2 +3 y +4 luego de la letra son numeros con un for y usando isdigit
                        #print("Acá hay una A")
                    
            #########################Funcion guardarAngulo
                    #poner rbt.RecibirOrden dentro de la clase robot
                                            # CAMBIAR W POR rbt.RecibirOrden()
   
    
    orden='EA+40B-30C180P'
    prueba=Trayectoria()
    prueba.esOrdenValida(orden)
    #Trayectoria.esOrdenValida
    print=("angulo 1 ", __self__.anguloGiro[0])
    print=("angulo 2 ", __self__.anguloGiro[1])
    print=("angulo 3 ", __self__.anguloGiro[2])

#si no anda el AnguloGiro usar ang1 ang2 y ang3 
    def setPosicionX(self, anguloGiro,):
            #Obtengo la posición en X del brazo
            #Primero paso el ángulo a radianes (por la librería)

        return(math.cos(math.radians(anguloGiro)))

    def setPosicionY(self, anguloGiro,):
            #Obtengo la posición en Y del brazo
            #Primero paso el ángulo a radianes (por la librería)

        return(math.cos(math.radians(anguloGiro)))   

    def setPosicionZ(self, anguloGiro,):
            #Obtengo la posición en Y del brazo
            #Primero paso el ángulo a radianes (por la librería)

        return(math.cos(math.radians(anguloGiro)))       
        
        #Acá abajo en vez de pasar una longitud estática se debe hacer con el atributo tamaño del robot
        #Lo hago con un def y un set_tamaño
        #También con un solo argumento ángulo alcanza. Se llama a los atributos de cada objeto
    def posicion(self, ang1, ang2, ang3):
            #longitudes de cada articulación
        l1=10 
        l2=5 
        l3=4.5
            #Posición del brazo A en el plano
        pos1x = 0
        pos1y = 0
        pos1z=0
            #Posición de la articulacion 1 en el plano
        pos2x = l2*Trayectoria.setPosicionX(ang2)
        pos2y = l2*Trayectoria.setPosicionY(ang2)
        pos2Z = l2*Trayectoria.setPosicionZ(ang2)
            #Posición de la articulacion 1 en el plano
        pos3x = l3*setPosicionX(ang3)+pos2x
        pos3y = l3*setPosicionX(ang3)+pos2y







    


